### PR TITLE
Do not automatically close XContent objects/arrays

### PR DESCRIPTION
**Pull request by tlrx**
Issue like #18433 appeared because [it's easy to forgot ](https://github.com/elastic/elasticsearch/commit/d7a31c8cf7f411aadc7202a4a2326679ed820e88) a `.endObject()` or `.endArray()` when building `XContent`objects using a `XContentBuilder`. 

And issues like this are even harder to detect because Jackson automatically closes unclosed objects and arrays in its default configuration, suggesting that all is well and the object has been correctly built while an object like 

``` json
{
  "a": {}, 
  "b": {}
}
```

should have been rendered like

``` json
{
  "a": {}, 
  "b": {
```

instead because of missing `.endObject()` calls.

This pull request disables the  `AUTO_CLOSE_JSON_CONTENT` feature so that json/cbor/smile objects are rendered _as is_ with missing `}` or `]` if necessary, and throws an exception if there are missing `endObject` or `endArray` calls when closing the `XContentBuilder`. It also fixes various tests that do not correctly close objects.

It is really a best effort to detect that `XContent`objects are not correctly built but it does not check every `ToXContent` object neither it replaces good REST tests.

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java [hunk 1]
```diff
@@ -41,6 +43,33 @@ public void testBasics() throws IOException {
...
+    }
+
+    public void testMissingEndArray() throws IOException {
+        try {
+            ByteArrayOutputStream os = new ByteArrayOutputStream();
+            try (XContentGenerator generator = xcontentType().xContent().createGenerator(os)) {
+                generator.writeStartArray();
+                generator.writeNumber(2L);
+            }
+            fail("should have thrown an exception concerning the unclosed array");
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java]: Can we convert this to the new `expectThrows` idiom?

File: core/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java [hunk 2]
```diff
@@ -41,6 +43,33 @@ public void testBasics() throws IOException {
...
 
+    public void testMissingEndObject() throws IOException {
+        try {
+            ByteArrayOutputStream os = new ByteArrayOutputStream();
+            try (XContentGenerator generator = xcontentType().xContent().createGenerator(os)) {
+                generator.writeStartObject();
+                generator.writeFieldName("foo");
+                generator.writeNumber(2L);
+            }
+            fail("should have thrown an exception concerning the unclosed object");
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java]: Can we convert this to the new `expectThrows` idiom?

File: core/src/test/java/org/elasticsearch/common/xcontent/builder/XContentBuilderTests.java [hunk 3]
```diff
@@ -363,4 +363,32 @@ public void testWriteFieldMapWithNullKeys() throws IOException {
...
         }
     }
+
+    public void testMissingEndObject() throws IOException {
+        try {
+            try (XContentBuilder builder = XContentFactory.contentBuilder(randomFrom(XContentType.values()))) {
+                builder.startObject();
+                builder.field("foo", true);
+            }
+            fail("should have thrown an exception concerning the unclosed object");
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/common/xcontent/builder/XContentBuilderTests.java]: Can we convert this to the new `expectThrows` idiom?

File: core/src/test/java/org/elasticsearch/common/xcontent/builder/XContentBuilderTests.java [hunk 4]
```diff
@@ -363,4 +363,32 @@ public void testWriteFieldMapWithNullKeys() throws IOException {
...
+
+    public void testMissingEndArray() throws IOException {
+        try {
+            try (XContentBuilder builder = XContentFactory.contentBuilder(randomFrom(XContentType.values()))) {
+                builder.startObject();
+                builder.startArray("foo");
+                builder.value(0);
+                builder.value(1);
+            }
+            fail("should have thrown an exception concerning the unclosed array");
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/common/xcontent/builder/XContentBuilderTests.java]: Can we convert this to the new `expectThrows` idiom?

File: core/src/test/java/org/elasticsearch/search/aggregations/AggregatorParsingTests.java [hunk 5]
```diff
@@ -225,25 +225,26 @@ public void testTwoTypes() throws Exception {
...
-                .endObject()
-                .endObject()
-                .startObject("aggs") // 2nd "aggs": illegal
-                .startObject("tag_count2")
-                .startObject("cardinality")
-                .field("field", "tag")
-                .endObject()
-                .endObject()
-                .endObject()
+                    .startObject("by_date")
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/search/aggregations/AggregatorParsingTests.java]: Thank you for formatting these. It makes them so much easier to read.

**Comment by jasontedor** [comment]
@tlrx It looks great. Thank you so much for doing this! I left a few comments about using the `expectThrows` idiom on the new tests (no need to change the existing tests).

File: core/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java (hunk 2 above)
Comment (jpountz) [inline core/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java]: +1

Comment (tlrx) [inline core/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java]: Sure!

**Comment by tlrx** [comment]
Thanks @jasontedor. I just updated the code according to your comments.Would you please have another look?

**Comment by jasontedor** [comment]
LGTM.

**Comment by tlrx** [comment]
@jasontedor thanks!

**Pull request merged**